### PR TITLE
[Merged by Bors] - feat(Group): constructors for minimal group axioms

### PR DESCRIPTION
**Pull request by ChrisHughes24**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Group/Defs.lean [hunk 1]
```diff
@@ -1160,6 +1160,50 @@ theorem CommGroup.toGroup_injective {G : Type u} : Function.Injective (@CommGrou
 #align comm_group.to_group_injective CommGroup.toGroup_injective
 #align add_comm_group.to_add_group_injective AddCommGroup.toAddGroup_injective
 
+/-- Define a Group structure on a Type by proving `∀ a, 1 * a = a` and
+`∀ a, a⁻¹ * a = 1`.
+See note [reducible non-instances]. -/
+@[to_additive (attr := reducible)]
+def Group.ofLeftAxioms {G : Type u} [Mul G] [Inv G] [One G]
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Defs.lean:1167]: Can you add the `to_additive` docstring too?

I think it's probably worth a note like "note that this uses the default recursive implementation for `npow` and `zpow`" (and `nsmul`/`zsmul` for the additive version).

File: Mathlib/Algebra/Group/Defs.lean [hunk 2]
```diff
@@ -1160,6 +1160,61 @@ theorem CommGroup.toGroup_injective {G : Type u} : Function.Injective (@CommGrou
...
+        simp only [assoc]
+      _ = 1 := by
+        rw [mul_right_inv, mul_one, mul_right_inv]
+  { mul_assoc := assoc,
+    mul_one := mul_one,
+    mul_left_inv := mul_left_inv,
+    one_mul := fun a => by
+      rw [← mul_right_inv a, assoc, mul_left_inv, mul_one] }
+#print AddGroup.ofLeftAxioms
+#print Group.ofLeftAxioms
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Defs.lean:1217]: Thee should be removed

File: Mathlib/Algebra/Group/Defs.lean [hunk 3]
```diff
@@ -1160,6 +1160,61 @@ theorem CommGroup.toGroup_injective {G : Type u} : Function.Injective (@CommGrou
...
 #align add_comm_group.to_add_group_injective AddCommGroup.toAddGroup_injective
 
+/-- Define a `Group` structure on a Type by proving `∀ a, 1 * a = a` and
+`∀ a, a⁻¹ * a = 1`.
+Note that this uses the default definitions for `npow`, `zpow` and `div`.
+See note [reducible non-instances]. -/
+@[to_additive (attr := reducible)
+"Define an `AddGroup` structure on a Type by proving `∀ a, 0 + a = a` and
+`∀ a, -a + a = 0`.
+Note that this uses the default definitions for `nsmul`, `zsmul` and `neg`.
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Defs.lean:1170]:
```suggestion
Note that this uses the default definitions for `nsmul`, `zsmul` and `sub`.
```

File: Mathlib/Algebra/Group/Defs.lean [hunk 4]
```diff
@@ -1160,6 +1160,61 @@ theorem CommGroup.toGroup_injective {G : Type u} : Function.Injective (@CommGrou
...
+      rw [← mul_left_inv a, ← assoc, mul_right_inv a, one_mul] }
+
+/-- Define a `Group` structure on a Type by proving `∀ a, a * 1 = a` and
+`∀ a, a * a⁻¹ = 1`.
+Note that this uses the default definitions for `npow`, `zpow` and `div`.
+See note [reducible non-instances]. -/
+@[to_additive (attr := reducible)
+"Define an `AddGroup` structure on a Type by proving `∀ a, a + 0 = a` and
+`∀ a, a + -a = 0`.
+Note that this uses the default definitions for `nsmul`, `zsmul` and `neg`.
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Defs.lean:1197]:
```suggestion
Note that this uses the default definitions for `nsmul`, `zsmul` and `sub`.
```

**Comment by ChrisHughes24** [comment]
I think it's worth spending a bit of time actually to get this right. Probably I should include `npow, `div` etc. as an `OptParam`. I'm not sure about this though. The difference should be invisible to most users right?

**Review by eric-wieser — APPROVED** [review]
LGTM, thanks! Will let someone else confirm that we do want these (and that they belong in this file and not `Group/ReducedAxioms` or something).

**Comment by eric-wieser** [comment]
> and not `Group/ReducedAxioms` or something

One argument for this is that if you also add `CommGroup` versions, we might have quite a lot of these, and `Group/Defs` is alraedy rather long.

**Comment by j-loreaux** [comment]
I would prefer they go in a new file just so that this `Algebra.Group.Defs` doesn't get even longer. It's only a `Defs` file after all! :smile: I don't care much about the name of said file.

**Review by eric-wieser — APPROVED** [review]
bors d+

Do you think it's worth adding a comment to the docstring of `Group` and `AddGroup` indicating that these helpers exist?

**Pull request closed without merging**